File: cv-generator-fe/src/api/user-memory/userMemory.ts
import { authFetch, readErrorMessage } from "../auth-utils/authFetch";

export type JobExperienceBullet = {
  id: string;
  bullet_points: string;
  relevant_technologies: string | null;
};

export type JobExperience = {
  id: string;
  company_name: string;
  job_title: string;
  start_date: string | null;
  end_date: string | null;
  location: string | null;
  bullets: JobExperienceBullet[];
};

export type EducationExperience = {
  id: string;
  degree: string;
  field_of_study: string | null;
  institution: string;
  start_date: string | null;
  end_date: string | null;
  description: string | null;
};

export type Project = {
  id: string;
  title: string;
  description: string | null;
  link: string | null;
};

export type Skill = {
  id: string;
  name: string;
};

export type Award = {
  id: string;
  title: string;
  issuer: string | null;
  date: string | null;
  description: string | null;
  link: string | null;
};

export type MemoryNote = {
  id: string;
  content: string;
};

export type UserMemory = {
  job_experiences: JobExperience[];
  education_experiences: EducationExperience[];
  projects: Project[];
  skills: Skill[];
  awards: Award[];
  notes: MemoryNote[];
};

export type IdPatch = {
  id?: string | null;
  delete?: boolean;
};

export type JobExperienceBulletPatch = IdPatch & Partial<Omit<JobExperienceBullet, "id">>;
export type JobExperiencePatch = IdPatch &
  Partial<Omit<JobExperience, "id" | "bullets">> & {
    bullets?: JobExperienceBulletPatch[];
  };
export type EducationExperiencePatch = IdPatch & Partial<Omit<EducationExperience, "id">>;
export type ProjectPatch = IdPatch & Partial<Omit<Project, "id">>;
export type SkillPatch = IdPatch & Partial<Omit<Skill, "id">>;
export type AwardPatch = IdPatch & Partial<Omit<Award, "id">>;
export type MemoryNotePatch = IdPatch & Partial<Omit<MemoryNote, "id">>;

export type UserMemoryPatch = {
  job_experiences?: JobExperiencePatch[];
  education_experiences?: EducationExperiencePatch[];
  projects?: ProjectPatch[];
  skills?: SkillPatch[];
  awards?: AwardPatch[];
  notes?: MemoryNotePatch[];
};

export async function getUserMemory(): Promise<UserMemory> {
  const response = await authFetch("/api/users/memory");

  if (!response.ok) {
    throw new Error(await readErrorMessage(response));
  }

  return (await response.json()) as UserMemory;
}

export async function updateUserMemory(patch: UserMemoryPatch): Promise<UserMemory> {
  const response = await authFetch("/api/users/memory", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(patch),
  });

  if (!response.ok) {
    throw new Error(await readErrorMessage(response));
  }

  return (await response.json()) as UserMemory;
}

File: cv-generator-be/app/schemas/user_memory.py
"""Request/response models for the user-memory editor API (`/users/memory`).

These are the CRUD DTOs the frontend memory editor speaks. They are distinct from
``schemas/memory.py``, which holds the LLM memory-*extraction* schemas.
"""

from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.config import MAX_MEMORY_NOTE_CHARS

# An incoming id may arrive as a UUID, a string (possibly blank), or be absent.
IdIn = UUID | str | None


class StrictModel(BaseModel):
    """Patch payloads reject unknown fields so typos surface as 422s."""

    model_config = ConfigDict(extra="forbid")


# ---------- output ----------
# ``from_attributes`` lets these be built straight from ORM rows via
# ``model_validate(row)``, so ``load_memory`` doesn't restate every field.


class OutModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class JobExperienceBulletOut(OutModel):
    id: UUID
    bullet_points: str
    relevant_technologies: str | None


class JobExperienceOut(OutModel):
    id: UUID
    company_name: str
    job_title: str
    start_date: str | None
    end_date: str | None
    location: str | None
    bullets: list[JobExperienceBulletOut]


class EducationExperienceOut(OutModel):
    id: UUID
    degree: str
    field_of_study: str | None
    institution: str
    start_date: str | None
    end_date: str | None
    description: str | None


class ProjectOut(OutModel):
    id: UUID
    title: str
    description: str | None
    link: str | None


class SkillOut(OutModel):
    id: UUID
    name: str


class AwardOut(OutModel):
    id: UUID
    title: str
    issuer: str | None
    date: str | None
    description: str | None
    link: str | None


class MemoryNoteOut(OutModel):
    id: UUID
    content: str


class UserMemoryOut(BaseModel):
    job_experiences: list[JobExperienceOut]
    education_experiences: list[EducationExperienceOut]
    projects: list[ProjectOut]
    skills: list[SkillOut]
    awards: list[AwardOut]
    notes: list[MemoryNoteOut]


# ---------- patch (upsert/delete) ----------


class JobExperienceBulletPatch(StrictModel):
    id: IdIn = None
    delete: bool = False
    bullet_points: str | None = None
    relevant_technologies: str | None = None


class JobExperiencePatch(StrictModel):
    id: IdIn = None
    delete: bool = False
    company_name: str | None = None
    job_title: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    location: str | None = None
    bullets: list[JobExperienceBulletPatch] | None = None


class EducationExperiencePatch(StrictModel):
    id: IdIn = None
    delete: bool = False
    degree: str | None = None
    field_of_study: str | None = None
    institution: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    description: str | None = None


class ProjectPatch(StrictModel):
    id: IdIn = None
    delete: bool = False
    title: str | None = None
    description: str | None = None
    link: str | None = None


class SkillPatch(StrictModel):
    id: IdIn = None
    delete: bool = False
    name: str | None = None


class AwardPatch(StrictModel):
    id: IdIn = None
    delete: bool = False
    title: str | None = None
    issuer: str | None = None
    date: str | None = None
    description: str | None = None
    link: str | None = None


class MemoryNotePatch(StrictModel):
    id: IdIn = None
    delete: bool = False
    content: str | None = Field(default=None, max_length=MAX_MEMORY_NOTE_CHARS)


class UserMemoryPatch(StrictModel):
    job_experiences: list[JobExperiencePatch] | None = None
    education_experiences: list[EducationExperiencePatch] | None = None
    projects: list[ProjectPatch] | None = None
    skills: list[SkillPatch] | None = None
    awards: list[AwardPatch] | None = None
    notes: list[MemoryNotePatch] | None = None
